jar-infer/jar-infer-lib/src/main/java/com/uber/nullaway/jarinfer/BytecodeAnnotator.java

```diff
@@ -46,77 +48,52 @@ private static void LOG(boolean cond, String tag, String msg) {
-  }
-
-  @Override
-  public void visit(
-      final int version,
-      final int access,
-      final String name,
-      final String signature,
-      final String superName,
-      final String[] interfaces) {
-    className = name.replace('/', '.');
-    if (cv != null) {
-      cv.visit(version, access, name, signature, superName, interfaces);
+  private static void addAnnotationIfNotPresent(
+      List<AnnotationNode> annotationList, String annotation) {
+    for (AnnotationNode node : annotationList) {
+      if (node.desc.equals(annotation)) {
+        return;
+      }
     }
+    annotationList.add(new AnnotationNode(Opcodes.ASM7, annotation));
   }
 
-  @Override
-  public MethodVisitor visitMethod(
-      final int access,
-      final String name,
-      final String desc,
-      final String signature,
-      final String[] exceptions) {
-    MethodVisitor mv = cv.visitMethod(access, name, desc, signature, exceptions);
-    LOG(debug, "DEBUG", "ClassName: " + className);
-    LOG(debug, "DEBUG", "visited method - " + name + " desc: " + desc + " sign: " + signature);
-    String methodSignature = className + "." + name + desc;
-    if (mv != null) {
+  private static void annotateBytecode(
+      InputStream is,
+      OutputStream os,
+      MethodParamAnnotations nonnullParams,
+      MethodReturnAnnotations nullableReturns)
+      throws IOException {
+    ClassReader cr = new ClassReader(is);
+    ClassWriter cw = new ClassWriter(0);
+    ClassNode cn = new ClassNode(Opcodes.ASM7);
+    cr.accept(cn, 0);
+
+    String className = cn.name.replace('/', '.');
+    List<MethodNode> methods = cn.methods;
+    for (MethodNode method : methods) {
+      String methodSignature = className + "." + method.name + method.desc;
       if (nullableReturns.contains(methodSignature)) {
         // Add a @Nullable annotation on this method to indicate that the method can return null.
-        mv.visitAnnotation(javaxNullableDesc, true);
+        addAnnotationIfNotPresent(method.visibleAnnotations, javaxNullableDesc);
         LOG(debug, "DEBUG", "Added nullable return annotation for " + methodSignature);
       }
-      Set<Integer> params = nullableParams.get(methodSignature);
+      Set<Integer> params = nonnullParams.get(methodSignature);
       if (params != null) {
-        boolean isStatic = (access & Opcodes.ACC_STATIC) != 0;
+        boolean isStatic = (method.access & Opcodes.ACC_STATIC) != 0;
         for (Integer param : params) {
+          int paramNum = isStatic ? param : param - 1;
           // Add a @Nonnull annotation on this parameter.
-          mv.visitParameterAnnotation(isStatic ? param : param - 1, javaxNonnullDesc, true);
+          addAnnotationIfNotPresent(method.visibleParameterAnnotations[paramNum], javaxNonnullDesc);
           LOG(
               debug,
               "DEBUG",
               "Added nullable parameter annotation for #" + param + " in " + methodSignature);
```

Comment:
Ok, I think the current code is correct, iterating over `nonnullParams` instead of `nullableParams` (Did our test miss that before? Or am I reading this wrong and this wasn't a bug in the existing code?), but then we probably want to change the debug messaged to "Added nonull parameter annotation ..." too. Right?

Reply:
Oh, I see there wasn't a bug before, just a variable name issue. It was always representing `nonnullParams` as passed to `annotateBytecode`, just poorly named inside the visitor. Maybe this kind of complexity is a point in favor of just going with the tree approach vs the visitor. Your call.

Either way, let's fix the debug log line :)

Reply:
Yeah, this was just an incorrect variable name. Fixed the debug log line.